flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java

```diff
@@ -0,0 +1,659 @@
+			totalFlinkMemorySize.add(jvmMetaspaceSize));
+		return new JvmMetaspaceAndOverhead(jvmMetaspaceSize, jvmOverheadSize);
+	}
+
+	private static FlinkInternalMemory deriveInternalMemoryFromTotalFlinkMemory(
+		final Configuration config, final MemorySize totalFlinkMemorySize) {
+		final MemorySize frameworkHeapMemorySize = getFrameworkHeapMemorySize(config);
+		final MemorySize taskOffHeapMemorySize = getTaskOffHeapMemorySize(config);
+
+		final MemorySize taskHeapMemorySize;
+		final MemorySize shuffleMemorySize;
+		final MemorySize managedMemorySize;
+
+		if (isTaskHeapMemorySizeExplicitlyConfigured(config)) {
+			// task heap memory is configured,
+			// derive managed memory first, leave the remaining to shuffle memory and check against shuffle min/max
+			taskHeapMemorySize = getTaskHeapMemorySize(config);
+			managedMemorySize = deriveManagedMemoryAbsoluteOrWithFraction(config, totalFlinkMemorySize);
+			final MemorySize totalFlinkExcludeShuffleMemorySize =
+				frameworkHeapMemorySize.add(taskHeapMemorySize).add(taskOffHeapMemorySize).add(managedMemorySize);
+			if (totalFlinkExcludeShuffleMemorySize.getBytes() > totalFlinkMemorySize.getBytes()) {
+				throw new IllegalConfigurationException(
+					"Sum of configured Framework Heap Memory (" + frameworkHeapMemorySize.toString()
+						+ "), Task Heap Memory (" + taskHeapMemorySize.toString()
+						+ "), Task Off-Heap Memory (" + taskOffHeapMemorySize.toString()
+						+ ") and Managed Memory (" + managedMemorySize.toString()
+						+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
+			}
+			shuffleMemorySize = totalFlinkMemorySize.subtract(totalFlinkExcludeShuffleMemorySize);
+			sanityCheckShuffleMemory(config, shuffleMemorySize, totalFlinkMemorySize);
+		} else {
+			// task heap memory is not configured
+			// derive managed memory and shuffle memory, leave the remaining to task heap memory
+			if (isManagedMemorySizeExplicitlyConfigured(config)) {
+				managedMemorySize = getManagedMemorySize(config);
+			} else {
+				managedMemorySize = deriveManagedMemoryAbsoluteOrWithFraction(config, totalFlinkMemorySize);
+			}
+			shuffleMemorySize = deriveShuffleMemoryWithFraction(config, totalFlinkMemorySize);
+			sanityCheckShuffleMemory(config, shuffleMemorySize, totalFlinkMemorySize);
+			final MemorySize totalFlinkExcludeTaskHeapMemorySize =
+				frameworkHeapMemorySize.add(taskOffHeapMemorySize).add(managedMemorySize).add(shuffleMemorySize);
+			if (totalFlinkExcludeTaskHeapMemorySize.getBytes() > totalFlinkMemorySize.getBytes()) {
+				throw new IllegalConfigurationException(
+					"Sum of configured Framework Heap Memory (" + frameworkHeapMemorySize.toString()
+						+ "), Task Off-Heap Memory (" + taskOffHeapMemorySize.toString()
+						+ "), Managed Memory (" + managedMemorySize.toString()
+						+ ") and Shuffle Memory (" + shuffleMemorySize.toString()
+						+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
```

Comment:
Without this check, it will fail at the next line `taskHeapMemorySize = totalFlinkMemorySize.subtract(totalFlinkExcludeTaskHeapMemorySize);`, because memory size cannot be negative. In that case, we may fail to provide helpful error message.

Reply:
Makes sense. Then let's keep it like this.